text/0000-variadic-tuples.md

```diff
@@ -0,0 +1,1006 @@
+- Feature Name: variadic_tuples
+- Start Date: 2019-08-16
+- RFC PR: [rust-lang/rfcs#0000](https://github.com/rust-lang/rfcs/pull/0000)
+- Rust Issue: [rust-lang/rust#0000](https://github.com/rust-lang/rust/issues/0000)
+
+# Summary
+
+[summary]: #summary
+
+Tuples types are ordered set of type, but users can only use tuple with fixed arity.
+
+This RFC aims to allow the use of a _variadic tuple_ which is a tuple with an arbitrary arity.
+
+# Motivation
+
+[motivation]: #motivation
+
+## Arbitrary tuple arity support
+
+Currently, when a user wants to either use or add behavior to tuples, he writes an impl for each tuple arity.
+For easier maintenance, it is usually done with a `macro_rules` and implemented up to 12 arity tuple. (see [ `Hash` implementation in `std`](https://github.com/rust-lang/rust/blob/master/src/libcore/hash/mod.rs)).
+
+Variadic tuple will provide several benefits considering trait implementation for tuple or using variadic tuples:
+
+- Implementations will be easier to write
+- Implementations will be easier to read and maintain
+- The compiler will compile implementation only for required tuple arity
+
+# Guide-level explanation
+
+[guide-level-explanation]: #guide-level-explanation
+
+Let's call a _variadic tuple type_ a tuple type with an arbitrary arity and a _variadic tuple_ an instance of a variadic tuple type.
+
+A variadic tuple type is declared with `(..T)` and a variadic tuple type can be expanded with `(..Vec<T>)`.
+
+Note: To illustrate the RFC, we will use the current implementation of the `Hash` trait for tuples.
+
+```rust
+// Quote from Rust source code
+// This macro implements `Hash` for a tuple.
+// It is used like this: `impl_hash_tuple! { A B C D E F }` for a 6-arity tuple.
+macro_rules! impl_hash_tuple {
+    () => (
+        #[stable(feature = "rust1", since = "1.0.0")]
+        impl Hash for () {
+            fn hash<H: Hasher>(&self, _state: &mut H) {}
+        }
+    );
+
+    ( $($name:ident)+) => (
+        #[stable(feature = "rust1", since = "1.0.0")]
+        impl<$($name: Hash),+> Hash for ($($name,)+) where last_type!($($name,)+): ?Sized {
+            #[allow(non_snake_case)]
+            fn hash<S: Hasher>(&self, state: &mut S) {
+                let ($(ref $name,)+) = *self;
+                $($name.hash(state);)+
+            }
+        }
+    );
+}
+
+macro_rules! last_type {
+    ($a:ident,) => { $a };
+    ($a:ident, $($rest_a:ident,)+) => { last_type!($($rest_a,)+) };
+}
+```
```

Comment:
Honestly? This should go in the motivation section. Show this ugly syntax, and compare it to the new one. It shouldn't go in the guide-level explanation, because this is just about explaining the new feature, not what it's replacing.

Reply:
I moved it to the motivation section and added other samples that can be interesting with this feature.